akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +210,166 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
 /**
  * INTERNAL API
  */
+final class GroupBy[T, K](maxSubstreams: Int, f: T ⇒ K) extends GraphStage[FlowShape[T, Source[T, NotUsed]]] {
+  val in: Inlet[T] = Inlet("GroupBy.in")
+  val out: Outlet[Source[T, NotUsed]] = Outlet("GroupBy.out")
+  override val shape: FlowShape[T, Source[T, NotUsed]] = FlowShape(in, out)
+  override def initialAttributes = DefaultAttributes.groupBy
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) with OutHandler with InHandler {
+    parent ⇒
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    private val activeSubstreamsMap = mutable.Map.empty[Any, SubstreamSource]
+    private val closedSubstreams = mutable.Set[Any]()
+    private var timeout: FiniteDuration = _
+    private var substreamWaitingToBePushed: Option[SubstreamSource] = None
+    private var nextElement: (K, T) = null
+    private var _nextId = 0
+
+    private def nextId(): Long = { _nextId += 1; _nextId }
+
+    private def hasNextElement = nextElement ne null
+
+    private def completeWholeStage(): Boolean =
+      if (activeSubstreamsMap.isEmpty || (!hasNextElement && !activeSubstreamsMap.values.exists(_.firstPush))) {
+        activeSubstreamsMap.values.foreach(_.complete())
+        completeStage()
+        true
+      } else false
+
+    private def fail(ex: Throwable): Unit = {
+      activeSubstreamsMap.values.foreach(_.fail(ex))
+      failStage(ex)
+    }
+
+    private def needToPull: Boolean = !(hasBeenPulled(in) || isClosed(in) || hasNextElement)
+
+    override def preStart(): Unit = timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+
+    override def onPull(): Unit = {
+      substreamWaitingToBePushed match {
+        case Some(substreamSource) ⇒
+          push(out, Source.fromGraph(substreamSource.source))
+          scheduleOnce(substreamSource.key, timeout)
+          substreamWaitingToBePushed = None
+        case None ⇒
+          if (hasNextElement) {
+            activeSubstreamsMap(nextElement._1).push(nextElement._2)
+            nextElement = null
+          } else tryPull(in)
+      }
+    }
+
+    override def onUpstreamFailure(ex: Throwable): Unit = fail(ex)
+
+    override def onDownstreamFinish(): Unit =
+      if (activeSubstreamsMap.isEmpty) completeStage() else setKeepGoing(true)
```

Comment:
is this always true? What if all active substreams have here demand pending, but since we have no pending elements (well, this needs to be checked) we should rather complete them?